eth_utils/abi.py

```diff
@@ -42,27 +83,682 @@ def collapse_if_tuple(abi: Dict[str, Any]) -> str:
+    else:
+        num_dims = tuple_dims.count("[")
+
+        # Arg is list tuple.  A non-list version of its abi will be used to
+        # align each element in `arg`.
+        new_abi = copy.copy(arg_abi)
+        new_abi["type"] = tuple_prefix + "[]" * (num_dims - 1)
+
+        sub_abis = itertools.repeat(new_abi)
+
+    if isinstance(arg, abc.Mapping):
+        # Arg is mapping.  Align values according to abi order.
+        aligned_arg = tuple(arg[abi["name"]] for abi in sub_abis)
+    else:
+        aligned_arg = arg
+
+    if not is_list_like(aligned_arg):
+        raise TypeError(
+            f'Expected non-string sequence for "{arg_abi.get("type")}" '
+            f"component type: got {aligned_arg}"
+        )
+
+    # convert NamedTuple to regular tuple
+    typing = tuple if isinstance(aligned_arg, tuple) else type(aligned_arg)
+
+    return typing(
+        _align_abi_input(sub_abi, sub_arg)
+        for sub_abi, sub_arg in zip(sub_abis, aligned_arg)
+    )
+
+
+def _get_tuple_type_str_and_dims(s: str) -> Optional[Tuple[str, Optional[str]]]:
+    """
+    Takes a JSON ABI type string.  For tuple type strings, returns the separated
+    prefix and array dimension parts.  For all other strings, returns ``None``.
+    """
+    tuple_type_str_re = "^(tuple)((\\[([1-9]\\d*\b)?])*)??$"
+    match = re.compile(tuple_type_str_re).match(s)
+
+    if match is not None:
+        tuple_prefix = match.group(1)
+        tuple_dims = match.group(2)
+
+        return tuple_prefix, tuple_dims
+
+    return None
+
+
+def abi_to_signature(abi: ABIElement) -> str:
+    """
+    Returns a string signature representation of the function or event ABI
+    and arguments.
+
+    Signatures consist of the name followed by a list of arguments.
+
+    :param abi: Contract ABI.
+    :type abi: `ABI`
+    :return: Stringified ABI signature
+    :rtype: `str`
+
+    .. doctest::
+
+        >>> from eth_utils import abi_to_signature
+        >>> abi = {
+        ...   'constant': False,
+        ...   'inputs': [
+        ...     {
+        ...       'name': 's',
+        ...       'type': 'uint256'
+        ...     }
+        ...   ],
+        ...   'name': 'f',
+        ...   'outputs': [],
+        ...   'payable': False,
+        ...   'stateMutability': 'nonpayable',
+        ...   'type': 'function'
+        ... }
+        >>> abi_to_signature(abi)
+        'f(uint256)'
+    """
+    if abi["type"] == "fallback" or abi["type"] == "receive":
+        name = str(abi["type"])
+        abi_inputs = None
+    elif abi["type"] == "constructor":
+        name = abi["type"]
+        abi_inputs = abi["inputs"]
+    elif abi["type"] == "event":
+        name = abi["name"]
+        abi_inputs = abi["inputs"]
+    else:
+        name = str(abi.get("name", abi["type"]))
+        abi_inputs = abi.get("inputs")
+
+    return f"{name}({_abi_inputs_types(abi_inputs)})"
+
+
+def filter_by_type(type: str, contract_abi: ABI) -> List[ABIElement]:
```

Comment:
This is the only method here that doesn't have `abi` in the name and I think it helps even though we're importing from `eth_utils.abi`. I also commented in the *web3.py* PR that we should move a very similar method here, defined there as `get_abi_by_name`. I think that would make a lot of sense to exist here and then rename this to match that name by adding "abi" to the method name.

This method signature re-enforces my other nits that `contract_abi` makes more sense as the arg names :)